common/utils.c

```diff
@@ -965,3 +984,257 @@ TDNFDirName(
+        dwError = ERROR_TDNF_INVALID_PARAMETER;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    ctx = EVP_MD_CTX_create();
+    if (!ctx)
+    {
+        pr_err("Context Create Failed\n");
+        dwError = ERROR_TDNF_CHECKSUM_VALIDATION_FAILED;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    dwError = EVP_DigestInit_ex(ctx, digest_type, NULL);
+    if (!dwError)
+    {
+        pr_err("Digest Init Failed\n");
+        dwError = ERROR_TDNF_CHECKSUM_VALIDATION_FAILED;
+        /* MD5 is not approved in FIPS mode. So, overrriding
+           the dwError to show the right error to the user */
+#if defined(OPENSSL_VERSION_MAJOR) && (OPENSSL_VERSION_MAJOR >= 3)
+        if (EVP_default_properties_is_fips_enabled(NULL) && !strcasecmp(hash->hash_type, "md5"))
+#else
+        if (FIPS_mode() && !strcasecmp(hash->hash_type, "md5"))
+#endif
+        {
+            dwError = ERROR_TDNF_FIPS_MODE_FORBIDDEN;
+        }
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    while ((length = read(fd, buf, (sizeof(buf)-1))) > 0)
+    {
+        dwError = EVP_DigestUpdate(ctx, buf, length);
+        if (!dwError)
+        {
+            pr_err("Digest Update Failed\n");
+            dwError = ERROR_TDNF_CHECKSUM_VALIDATION_FAILED;
+            BAIL_ON_TDNF_ERROR(dwError);
+        }
+        memset(buf, 0, BUFSIZ);
+    }
+
+    if (length == -1)
+    {
+        pr_err("Metalink: validating (%s) FAILED\n", filename);
+        dwError = errno;
+        BAIL_ON_TDNF_SYSTEM_ERROR(dwError);
+    }
+
+    dwError = EVP_DigestFinal_ex(ctx, digest, &digest_length);
+    if (!dwError)
+    {
+        pr_err("Digest Final Failed\n");
+        dwError = ERROR_TDNF_CHECKSUM_VALIDATION_FAILED;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+    dwError = 0;
+
+cleanup:
+    if (fd >= 0)
+    {
+        close(fd);
+    }
+    if (ctx)
+    {
+        EVP_MD_CTX_destroy(ctx);
+    }
+    return dwError;
+error:
+    goto cleanup;
+}
+
+uint32_t
+TDNFCheckHash(
+    const char *filename,
+    unsigned char *digest,
+    int type
+    )
+{
+
+    uint32_t dwError = 0;
+    uint8_t digest_from_file[EVP_MAX_MD_SIZE] = {0};
+    hash_op *hash = NULL;
+
+    if (IsNullOrEmptyString(filename) ||
+       !digest)
+    {
+        dwError = ERROR_TDNF_INVALID_PARAMETER;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    if (type  < TDNF_HASH_MD5 || type >= TDNF_HASH_SENTINEL)
+    {
+        dwError = ERROR_TDNF_INVALID_PARAMETER;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    hash = hash_ops + type;
+
+    dwError = TDNFGetDigestForFile(filename, hash, digest_from_file);
+    BAIL_ON_TDNF_ERROR(dwError);
+
+    if (memcmp(digest_from_file, digest, hash->length))
+    {
+        dwError = ERROR_TDNF_CHECKSUM_VALIDATION_FAILED;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+cleanup:
+    return dwError;
+error:
+    if (!IsNullOrEmptyString(filename))
+    {
+        pr_err("Error: Validating metalink (%s) FAILED (digest mismatch)\n", filename);
```

Comment:
You need to remove references to "metalink" since the code isn't specific to it any more.